pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/properties/PreserveStashesJobProperty.java

```diff
@@ -0,0 +1,122 @@
+import org.jenkinsci.plugins.workflow.job.WorkflowJob;
+import org.jenkinsci.plugins.workflow.job.WorkflowRun;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.DataBoundSetter;
+import org.kohsuke.stapler.QueryParameter;
+
+import javax.annotation.Nonnull;
+import java.io.IOException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public class PreserveStashesJobProperty extends OptionalJobProperty<WorkflowJob> {
+
+    private static final Logger LOGGER = Logger.getLogger(PreserveStashesJobProperty.class.getName());
+
+    private int buildCount = 10;
+
+    @DataBoundConstructor
+    public PreserveStashesJobProperty() {
+
+    }
+
+    public int getBuildCount() {
+        return buildCount;
+    }
+
+    @DataBoundSetter
+    public void setBuildCount(int buildCount) {
+        this.buildCount = buildCount;
+    }
+
+    @Extension @Symbol("preserveStashes")
+    public static class DescriptorImpl extends OptionalJobPropertyDescriptor {
+        public FormValidation doCheckBuildCount(@QueryParameter int value) {
+            if (value < 1) {
+                return FormValidation.error("Must be greater than 0");
+            } else {
+                return FormValidation.ok();
+            }
+        }
+    }
+
+    @Extension
+    public static final class SaveStashes extends StashManager.StashBehavior {
+
+        @Override
+        public boolean shouldClearAll(@Nonnull Run<?,?> build) {
+            if (build instanceof WorkflowRun) {
+                WorkflowRun r = (WorkflowRun)build;
+                WorkflowJob j = r.getParent();
+                PreserveStashesJobProperty prop = j.getProperty(PreserveStashesJobProperty.class);
+                if (prop != null) {
+                    int bc = prop.getBuildCount();
+                    // Being particularly defensive just to be safe.
+                    if (bc > 0) {
+                        for (WorkflowRun recentRun : j.getBuilds().completedOnly().limit(bc)) {
+                            if (recentRun != null && recentRun.equals(r)) {
+                                return false;
+                            }
+                        }
+                    }
+                }
+            }
+            return true;
+        }
+    }
+
+    @Extension
+    public static final class StashClearingListener extends RunListener<Run<?,?>> {
+        @Override
+        public void onCompleted(Run<?,?> r, TaskListener listener) {
+            if (r instanceof WorkflowRun) {
+                WorkflowJob j = ((WorkflowRun) r).getParent();
+                PreserveStashesJobProperty prop = j.getProperty(PreserveStashesJobProperty.class);
+                if (prop != null && prop.getBuildCount() > 0) {
+                    // TODO: Determine if this is overzealous. Might need to ape log rotation more closely.
+                    for (WorkflowRun build : j.getBuilds().completedOnly()) {
```

Comment:
This will be super inefficient for jobs with a lot of builds. You need to use `RunLoadCounter` to verify that you are not destroying performance.

Reply:
Will do.

Reply:
Blech. Finally got around to looking into this and realized `RunLoadCounter` is `AbstractProject`-only. Need to think about how to rework it to be Pipeline-friendly.

Reply:
and got it.

Reply:
So what was the result? I feel like the test may be wrong, since I do not see how this could _not_ load too much.

Reply:
Yeah, I had the same suspicion. Which is why I put the `lastFailed` check in the test too - so it looks, at the very least, like the `completedOnly()` check doesn't actually lead to loading actions. I'm still wary.

Reply:
I am fairly confident this is still wrong, and if your test is not catching it, you got the test wrong somehow.

Worse, for `artifact-manager-s3`, by default this call will do nothing, so loading build records to do nothing is a pure waste.

Reply:
fwiw, in debug testing, the test only calls `WorkflowRun#onLoad` once total, for build 10, the most recent build as of the restart.

Reply:
> You mean `StashManager.maybeClearAll(build)` won't do anything for `artifact-manager-s3`?

By default, yes.

> something is definitely different between those cases

Sounds like something is wrong with the test. Debug it.

> I don't see how this'd be any worse than, say, `LogRotator`

Because `LogRotator` _deletes_ builds that are too old, so the next time it runs, it does not need to iterate them.

> `LogRotator` is looking at more data for each run than this is, since it's looking at `r.getTimestamp()` and `r.isKeepLog()`

Mostly irrelevant. The expensive part is loading the `Run` object from `build.xml`.

> the stash clearing stuff literally just looks for completed builds

Of which there could be a million. Every one of which you will iterate, forcing it to be loaded into heap, every time a build completes.

`SaveStashes#shouldClearAll` looks OK, because it uses a `limit`.

Reply:
I did debug it - as I said, `onLoad` is only called once. 

But ok, I'll throw a limit on the maximum number of builds-with-stashes to keep, and then set a limit in `StashClearingListener#onCompleted` to go with it.